smart-contracts/contracts/burn_coins/src/contract.rs

```diff
@@ -0,0 +1,140 @@
+use cosmwasm_std::{
+    entry_point, to_json_binary, BankMsg, Binary, CosmosMsg, Deps, DepsMut, Env, MessageInfo,
+    Response, StdResult,
+};
+use cw2::set_contract_version;
+use cw20_base::msg::MigrateMsg;
+
+use crate::msg::{ExecuteMsg, InstantiateMsg, QueryMsg};
+use crate::query::query_total_burn;
+use crate::state::AMOUNT_BURNT;
+use crate::BurnErrors;
+
+// version info for migration info
+const CONTRACT_NAME: &str = "burn_coins";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    _msg: InstantiateMsg,
+) -> Result<Response, BurnErrors> {
+    // Set the contract version in storage
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    // Return a response
+    Ok(Response::new())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, BurnErrors> {
+    match msg {
+        ExecuteMsg::Burn {} => execute_coins_burn(deps, info),
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::TotalBurntQuery {} => to_json_binary(&query_total_burn(deps)?),
+    }
+}
+
+#[entry_point]
+pub fn migrate(_deps: DepsMut, _env: Env, _msg: MigrateMsg) -> Result<Response, BurnErrors> {
+    Ok(Response::new().add_attribute("migrate", "successful"))
+}
+
+pub fn execute_coins_burn(deps: DepsMut, info: MessageInfo) -> Result<Response, BurnErrors> {
+    if info.funds.is_empty() {
+        return Err(BurnErrors::ZeroAmount {});
+    }
+
+    // Prepare the Burn message
+    let burn_msg = CosmosMsg::Bank(BankMsg::Burn {
+        amount: info.clone().funds,
+    });
+
+    for fund in info.funds {
+        if let Some(mut total_burn_amount) =
+            AMOUNT_BURNT.may_load(deps.storage, fund.clone().denom)?
+        {
+            total_burn_amount += fund.clone().amount;
+            AMOUNT_BURNT.save(deps.storage, fund.clone().denom, &total_burn_amount)?;
+        } else {
+            AMOUNT_BURNT.save(deps.storage, fund.clone().denom, &fund.clone().amount)?;
+        }
+    }
+
+    // Return a response with the burn message
+    Ok(Response::default().add_message(burn_msg))
```

Comment:
This looks indeed correct. But im just wondering in the case we want to build a Subgraph indexer, would we need some extra attribute here? If any it should be something very minimalistic, but if too much minimalistic it could be useless. So this could be probably skipped in favor of an hardcoded contract addy on the subgraph side, looking for sdk.Burn events.

What do you think?

Reply:
This can be indexed by checking the bank module burn event. That will be a more reliable event to index.